### PR TITLE
New backups (15): Put together backup creation in the use case

### PR DESCRIPTION
**Pull request by makingthematrix**
#### APK
[Download build #2548](http://10.10.124.11:8080/job/Pull%20Request%20Builder/2548/artifact/build/artifact/wire-dev-PR2982-2548.apk)
[Download build #2550](http://10.10.124.11:8080/job/Pull%20Request%20Builder/2550/artifact/build/artifact/wire-dev-PR2982-2550.apk)
[Download build #2554](http://10.10.124.11:8080/job/Pull%20Request%20Builder/2554/artifact/build/artifact/wire-dev-PR2982-2554.apk)

### CODE REVIEW

File: app/src/main/kotlin/com/waz/zclient/feature/backup/usecase/CreateBackUpUseCase.kt [hunk 1]
```diff
@@ -1,33 +1,66 @@
...
+import java.text.SimpleDateFormat
+import java.util.Locale
 
 class CreateBackUpUseCase(
     private val backUpRepositories: List<BackUpRepository<File>>,
+    private val zipHandler: ZipHandler,
+    private val encryptionHandler: EncryptionHandler,
     private val coroutineScope: CoroutineScope = CoroutineScope(SupervisorJob() + Dispatchers.Default)
-) : UseCase<Unit, Unit> {
+) : UseCase<File, Triple<UserId, String, String>> {
```
Comment (gizemb) [inline app/src/main/kotlin/com/waz/zclient/feature/backup/usecase/CreateBackUpUseCase.kt:27]: i think it's better to create an encapsulating class instead of using `Triple` here. it's hard to follow what the parameters are in that way and it's error prone. i'd be more comfortable with a class that has `userId`, `userHandle` and `password` parameters. then you wouldn't have to do lines 30-32 below.

File: app/src/main/kotlin/com/waz/zclient/feature/backup/usecase/CreateBackUpUseCase.kt [hunk 2]
```diff
@@ -1,33 +1,66 @@
...
+    private fun backupZipFileName(userHandle: String): String {
+        val timestamp = SimpleDateFormat("yyyyMMdd", Locale.getDefault()).format(Instant.now().epochSecond * 1000)
+        return "Wire-$userHandle-Backup_$timestamp.android_wbu"
+    }
+
+    private fun extractFiles(list: List<Either<Failure, File>>): Either<Failure, List<File>> {
+        val files = mutableListOf<File>()
 
-object BackUpCreationFailure : FeatureFailure()
+        for (value in list)
```
Comment (gizemb) [inline app/src/main/kotlin/com/waz/zclient/feature/backup/usecase/CreateBackUpUseCase.kt:54]: i'd prefer having brackets for `for` loop here since it has a multi line operation.

File: app/src/main/kotlin/com/waz/zclient/feature/backup/usecase/CreateBackUpUseCase.kt [hunk 3]
```diff
@@ -1,33 +1,66 @@
...
+            when (value) {
+                is Either.Right -> files += value.b
+                is Either.Left -> return Either.Left(value.a)
+            }
+
+        return Either.Right(files.toList())
```
Comment (gizemb) [inline app/src/main/kotlin/com/waz/zclient/feature/backup/usecase/CreateBackUpUseCase.kt:60]: why do we return `Either.Right` if there's a `Failure` along the way? then getting `Either` as a parameter doesn't mean anything if we simply ignore them.

File: app/src/main/kotlin/com/waz/zclient/feature/backup/usecase/CreateBackUpUseCase.kt [hunk 4]
```diff
@@ -1,33 +1,66 @@
...
-}
+    )
+
+    @SuppressWarnings("MagicNumber")
+    private fun backupZipFileName(userHandle: String): String {
+        val timestamp = SimpleDateFormat("yyyyMMdd", Locale.getDefault()).format(Instant.now().epochSecond * 1000)
+        return "Wire-$userHandle-Backup_$timestamp.android_wbu"
+    }
+
+    private fun extractFiles(list: List<Either<Failure, File>>): Either<Failure, List<File>> {
```
Comment (gizemb) [inline app/src/main/kotlin/com/waz/zclient/feature/backup/usecase/CreateBackUpUseCase.kt:51]: what is a `list`? what does it contain? i think it needs a more meaningful name.

File: app/src/main/kotlin/com/waz/zclient/feature/backup/usecase/CreateBackUpUseCase.kt [hunk 5]
```diff
@@ -1,33 +1,66 @@
...
+        val files = mutableListOf<File>()
 
-object BackUpCreationFailure : FeatureFailure()
+        for (value in list)
+            when (value) {
+                is Either.Right -> files += value.b
+                is Either.Left -> return Either.Left(value.a)
+            }
```
Comment (gizemb) [inline app/src/main/kotlin/com/waz/zclient/feature/backup/usecase/CreateBackUpUseCase.kt:58]: `value.onSuccess { files += it }.onFailure{ ... }`

File: app/src/main/kotlin/com/waz/zclient/feature/backup/zip/ZipHandler.kt [hunk 6]
```diff
@@ -0,0 +1,10 @@
+package com.waz.zclient.feature.backup.zip
+
+import com.waz.zclient.core.exception.Failure
+import com.waz.zclient.core.functional.Either
+import java.io.File
+
+interface ZipHandler {
+    fun zip(zipFileName: String, files: List<File>): Either<Failure, File>
+    fun unzip(zipFile: File): Either<Failure, List<File>>
+}
```
Comment (gizemb) [inline app/src/main/kotlin/com/waz/zclient/feature/backup/zip/ZipHandler.kt:10]: i made a similar comment in `backup_16`. there's no benefit of extracting this to an interface if there's only one concrete class implementing it.

File: app/src/test/kotlin/com/waz/zclient/feature/backup/ZipHandlerTest.kt [hunk 7]
```diff
@@ -2,15 +2,15 @@ package com.waz.zclient.feature.backup
...
 import com.waz.zclient.feature.uniqueZipFileName
 import org.amshove.kluent.`should be greater than`
 import org.amshove.kluent.shouldEqual
 import org.junit.Assert.fail
 import org.junit.Test
 import java.io.File
-import java.util.Base64
-import kotlin.random.Random
 
 class ZipHandlerTest : UnitTest() {
```
Comment (gizemb) [inline app/src/test/kotlin/com/waz/zclient/feature/backup/ZipHandlerTest.kt:15]: if this only tests the `ZipHandlerDataSource`, then it's now a `ZipHandlerDataSourceTest`, not the test of the interface. the interface does not have any implementation to test after all.

File: app/src/test/kotlin/com/waz/zclient/feature/backup/usecase/CreateBackUpUseCaseTest.kt [hunk 8]
```diff
@@ -21,51 +31,130 @@ class CreateBackUpUseCaseTest : UnitTest() {
...
+            assertEquals(Either.Left(FakeEncryptionFailure), result)
         }
     }
 
     companion object {
+        /**
+         * Returns Mockito.any() as nullable type to avoid java.lang.IllegalStateException when null is returned.
+         * Taken from https://stackoverflow.com/a/48091649/2975925
+         */
+        fun <T> any(): T = Mockito.any<T>()
```
Comment (gizemb) [inline app/src/test/kotlin/com/waz/zclient/feature/backup/usecase/CreateBackUpUseCaseTest.kt:134]: we already have this in MockitoKotlin.kt. just use the correct import.

File: app/src/main/kotlin/com/waz/zclient/feature/backup/usecase/CreateBackUpUseCase.kt (hunk 1 above)
Comment (makingthematrix) [reply app/src/main/kotlin/com/waz/zclient/feature/backup/usecase/CreateBackUpUseCase.kt:27]: Right. Changed in https://github.com/wireapp/wire-android/pull/2988.

File: app/src/main/kotlin/com/waz/zclient/feature/backup/usecase/CreateBackUpUseCase.kt (hunk 3 above)
Comment (gizemb) [reply app/src/main/kotlin/com/waz/zclient/feature/backup/usecase/CreateBackUpUseCase.kt:60]: ok, so it returns the failure at line 57

File: app/src/main/kotlin/com/waz/zclient/feature/backup/usecase/CreateBackUpUseCase.kt (hunk 2 above)
Comment (makingthematrix) [reply app/src/main/kotlin/com/waz/zclient/feature/backup/usecase/CreateBackUpUseCase.kt:54]: Changed to `forEach` in https://github.com/wireapp/wire-android/pull/2988/commits/393aab2c7ba823d4609fe30d1549c26804049ce9

File: app/src/main/kotlin/com/waz/zclient/feature/backup/usecase/CreateBackUpUseCase.kt (hunk 4 above)
Comment (makingthematrix) [reply app/src/main/kotlin/com/waz/zclient/feature/backup/usecase/CreateBackUpUseCase.kt:51]: Change to `filesOrFailure` in https://github.com/wireapp/wire-android/pull/2988/commits/393aab2c7ba823d4609fe30d1549c26804049ce9

File: app/src/main/kotlin/com/waz/zclient/feature/backup/usecase/CreateBackUpUseCase.kt (hunk 5 above)
Comment (makingthematrix) [reply app/src/main/kotlin/com/waz/zclient/feature/backup/usecase/CreateBackUpUseCase.kt:58]: Not possible because of that early `return`...

File: app/src/main/kotlin/com/waz/zclient/feature/backup/zip/ZipHandler.kt (hunk 6 above)
Comment (makingthematrix) [reply app/src/main/kotlin/com/waz/zclient/feature/backup/zip/ZipHandler.kt:10]: Fixed in https://github.com/wireapp/wire-android/pull/2988/commits/393aab2c7ba823d4609fe30d1549c26804049ce9

File: app/src/test/kotlin/com/waz/zclient/feature/backup/ZipHandlerTest.kt (hunk 7 above)
Comment (makingthematrix) [reply app/src/test/kotlin/com/waz/zclient/feature/backup/ZipHandlerTest.kt:15]: `ZipHandlerDataSource` changed to `ZipHandlerTest` in https://github.com/wireapp/wire-android/pull/2988/commits/393aab2c7ba823d4609fe30d1549c26804049ce9

File: app/src/test/kotlin/com/waz/zclient/feature/backup/usecase/CreateBackUpUseCaseTest.kt (hunk 8 above)
Comment (makingthematrix) [reply app/src/test/kotlin/com/waz/zclient/feature/backup/usecase/CreateBackUpUseCaseTest.kt:134]: Changed in https://github.com/wireapp/wire-android/pull/2988/commits/393aab2c7ba823d4609fe30d1549c26804049ce9

File: app/src/main/kotlin/com/waz/zclient/feature/backup/usecase/CreateBackUpUseCase.kt [hunk 9]
```diff
@@ -1,33 +1,66 @@
...
             .map { coroutineScope.async(Dispatchers.IO) { it.saveBackup() } }
             .awaitAll()
-            .any { it.isLeft }
-}
+    )
+
+    @SuppressWarnings("MagicNumber")
+    private fun backupZipFileName(userHandle: String): String {
+        val timestamp = SimpleDateFormat("yyyyMMdd", Locale.getDefault()).format(Instant.now().epochSecond * 1000)
+        return "Wire-$userHandle-Backup_$timestamp.android_wbu"
```
Comment (gizemb) [inline app/src/main/kotlin/com/waz/zclient/feature/backup/usecase/CreateBackUpUseCase.kt:48]: this can be taken out to a separate class that handles date formatting. it'd be nice for reusability. also the date format is a magic string here.

i'll approve this PR now as the reviews are piling up, but i'd be glad if you could do this in a separate PR.

**Review by gizemb — APPROVED** [review]

File: app/src/main/kotlin/com/waz/zclient/feature/backup/usecase/CreateBackUpUseCase.kt (hunk 9 above)
Comment (makingthematrix) [reply app/src/main/kotlin/com/waz/zclient/feature/backup/usecase/CreateBackUpUseCase.kt:48]: @gizemb : I moved it to `DateAndTimeUtils` in a new PR: https://github.com/wireapp/wire-android/pull/3004
